collectors/fda_dap/spider.py

```diff
@@ -0,0 +1,266 @@
+            meta = {
+                'cookiejar': random.random(),
+            }
+            return Request(
+                url,
+                cookies=original_cookies,
+                meta=meta,
+                dont_filter=True,
+                callback=self.parse_drug_details_or_overview
+            )
+        else:
+            msg = 'Don\'t know how to handle URL (%s)' % response.url
+            raise Exception(msg)
+
+    def parse_drug_details(self, response):
+        '''
+        We're getting closer! We're already on the drug details page, now we
+        just need to look for the Approval History link (if it exists). The
+        state is still kept in the cookies, so we need to make sure we're using
+        the same CookieJar as the one used on the request here.
+        '''
+        url_candidates = response.css('#user_provided li > a::attr(href)').extract()
+        for url in url_candidates:
+            if 'Search.Label_ApprovalHistory' in url:
+                meta = {
+                    'cookiejar': response.meta.get('cookiejar'),
+                }
+                return Request(
+                    urlparse.urljoin(response.url, url),
+                    meta=meta,
+                    dont_filter=True,
+                    callback=self.parse_approval_history
+                )
+
+    def parse_drug_overview(self, response):
+        '''
+        This page contains multiple applications for a single drug. We need to
+        go to the Drug Details page of each one of those. When we click on a
+        link here, we're redirected to a page which sets the state (in the
+        cookies) to the drug's details, then redirecting us to the drug details
+        page. To avoid the requests messing up with each other, we need a clean
+        set of cookies for each one.
+        '''
+        urls = response.css('.product_table a::attr(href)').extract()
+        for url in urls:
+            meta = {
+                'cookiejar': random.random(),
+            }
+            yield Request(
+                urlparse.urljoin(response.url, url),
+                meta=meta,
+                dont_filter=True,
+                callback=self.parse_drug_details_or_overview
+            )
+
+    def _response_cookies_to_dict(self, response):
+        cookies = CookieJar().make_cookies(response, response.request)
+
+        def _cookie_to_dict(cookie):
+            return {
+                'name': cookie.name,
+                'value': cookie.value,
+                'domain': cookie.domain,
+                'path': cookie.path,
+            }
+
+        return [_cookie_to_dict(cookie) for cookie in cookies]
+
+    def parse_approval_history(self, response):
+        _extract_text = lambda selector: selector.xpath('text()') \
+                                                 .extract_first() \
+                                                 .strip()
+
+        def _parse_links(selector):
```

Comment:
It doesn't use anything from the closure, so could be extracted. I wrote it here as I was developing, but didn't went through it to think about the best place to put it.